lib/collection/src/shard/replica_set.rs

```diff
@@ -18,23 +18,61 @@ pub type IsActive = bool;
 /// A set of shard replicas.
 /// Handles operations so that the state is consistent across all the replicas of the shard.
 pub struct ReplicaSet {
+    shard_id: ShardId,
     this_peer_id: PeerId,
     local: Option<LocalShard>,
     // TODO: Remote shard should be able to query several peers
     remote: Option<RemoteShard>,
-    replica_state: HashMap<PeerId, IsActive>,
+    pub(crate) replica_state: HashMap<PeerId, IsActive>,
 }
 
 impl ReplicaSet {
+    pub fn peer_ids(&self) -> Vec<PeerId> {
+        todo!()
+    }
+
     pub fn set_active(&mut self, peer_id: &PeerId, active: bool) -> CollectionResult<()> {
         *self
             .replica_state
             .get_mut(peer_id)
             .ok_or_else(|| CollectionError::NotFound {
-                what: format!("Replica on peer {peer_id}"),
+                what: format!("Shard {} replica on peer {peer_id}", self.shard_id),
             })? = active;
         Ok(())
     }
+
+    pub async fn apply_state(
+        &mut self,
+        replicas: HashMap<PeerId, IsActive>,
+    ) -> CollectionResult<()> {
+        let removed_peers = self
+            .replica_state
+            .keys()
+            .filter(|peer_id| !replicas.contains_key(peer_id))
+            .copied()
+            .collect::<Vec<_>>();
+        for peer_id in removed_peers {
+            if peer_id == self.this_peer_id {
+                if let Some(shard) = &mut self.local {
```

Comment:
Is it possible to get a `None` here?

Reply:
It should not be possible generally, having it here would mean there was an error in logic somewhere

Reply:
What do you think about adding a debug assert?

Reply:
I'd like a debug assert please :pray: